pdns/recursordist/syncres.cc

```diff
@@ -4439,6 +4441,12 @@ void SyncRes::sanitizeRecordsPass2(const std::string& prefix, LWResult& lwr, con
         LOG(prefix << qname << ": Removing irrelevent record '" << rec->toString() << "' in the ANSWER section received from " << auth << endl);
         skipvec[counter] = true;
         ++skipCount;
+      }
+      // If we have a CNAME, skip answer records for the requested type
+      if (rec->d_type == qtype && rec->d_name == qname && cnamesSeen.find(rec->d_name) != cnamesSeen.end() && qtype != QType::CNAME) {
```

Comment:
Are you sure about the `rec->d_name == qname` part? It seems to defeat the carefully constructed list of names present in the CNAME chain?

Reply:
Hmm, indeed, as it is written now it will only work for the original qname. Will revisit.

Reply:
I added a test to check for the case of longer CNAME chains with answer records mixed in and it does seem to do the right thing already.

As we restart the resolve on the first CNAME to the target seen and we did not add the extra (CNAME or answer) records to the result to the answers when we do the restart. 

But please check my reasoning, I might be overlooking something.

Reply:
You are right, so perhaps we can replace `cnamesSeen` by a boolean indicating whether a `CNAME` for the `qname` has been seen?

Reply:
done